# Pre-processing

# Objective

- Prepare data for modelling

## Tasks

1. Find features with missing values
2. Determine relevance of feature for modelling. Drop features you feel are not important
3. Impute numerical missing values
4. Convert categorical variables and datetime variables to numerical variables
5. Impute categorical missing values
6. Scale all data

# Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score

# Load Data

In [2]:
read = pd.read_csv('..\data\interim\\rentals_complete.csv')

In [3]:
rentals = read.copy()

In [4]:
rentals.shape

(267690, 36)

In [5]:
rentals.head()

state  serviceCharge newlyConst balcony hasKitchen  baseRent  \
0  Nordrhein_Westfalen         245.00         No      No         No    595.00   
1  Nordrhein_Westfalen         141.00         No     Yes         No    579.00   
2  Nordrhein_Westfalen         141.00         No     Yes         No    569.00   
3  Nordrhein_Westfalen         322.16        Yes     Yes        Yes   1398.60   
4  Nordrhein_Westfalen         112.92         No      No        Yes    725.62   

   livingSpace       condition   interiorQual lift  ...  \
0        86.00       well_kept         normal   No  ...   
1        70.95         unknown        unknown   No  ...   
2        70.95         unknown        unknown   No  ...   
3       115.58  first_time_use  sophisticated  Yes  ...   
4        51.10  mint_condition  sophisticated  Yes  ...   

  medianServiceCharge_ct  medianThermalChar_ct  medianPictureCount_ct  \
0                  121.0                 126.0                    8.0   
1                  121.0                 126.0                    8.0   
2                  121.0                 126.0                    8.0   
3                  121.0                 126.0                    8.0   
4                  121.0                 126.0                    8.0   

  medianNoRooms_muni medianLivingSpace_muni medianServiceCharge_muni  \
0                3.0                   78.0                   161.62   
1                3.0                   78.0                   161.62   
2                3.0                   78.0                   161.62   
3                3.0                   78.0                   161.62   
4                3.0                   78.0                   161.62   

  medianPictureCount_muni  medianNoRooms_zip  medianLivingSpace_zip  \
0                     9.0                3.0                   78.0   
1                     9.0                3.0                   78.0   
2                     9.0                3.0                   78.0   
3                     9.0                3.0                   78.0   
4                     9.0                3.0                   78.0   

   medianServiceCharge_zip  
0                    150.0  
1                    150.0  
2                    150.0  
3                    150.0  
4                    150.0  

[5 rows x 36 columns]

In [6]:
rentals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 267690 entries, 0 to 267689
Data columns (total 36 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   state                      267690 non-null  object 
 1   serviceCharge              257456 non-null  float64
 2   newlyConst                 267690 non-null  object 
 3   balcony                    267690 non-null  object 
 4   hasKitchen                 267690 non-null  object 
 5   baseRent                   267670 non-null  float64
 6   livingSpace                267690 non-null  float64
 7   condition                  267690 non-null  object 
 8   interiorQual               267690 non-null  object 
 9   lift                       267690 non-null  object 
 10  typeOfFlat                 267690 non-null  object 
 11  zip_code                   267690 non-null  int64  
 12  noRooms                    267690 non-null  float64
 13  city_town                  26

# Reviewing missing values

In [7]:
#Check columns with missing values
rentals.columns[rentals.isnull().any()]

Index(['serviceCharge', 'baseRent', 'description', 'facilities',
       'medianServiceCharge_muni', 'medianServiceCharge_zip'],
      dtype='object')

We will impute median instead of mean values for the missing values for serviceCharge and and drop missing values for baseRent. We will impute mean values for the missing values for medianServiceCharge_muni and medianServiceCharge_zip because these values are unlikely to have outliers.

In [8]:
#Dropping baseRent values that are null
rentals = rentals[~(rentals.baseRent.isnull())]
#Replace missing values for serviceCharge and baseRent with median
rentals.loc[rentals.serviceCharge.isnull(), 'serviceCharge'] = rentals.serviceCharge.median()
#Replace missing values for medianServiceCharge_muni and medianServiceCharge_zip with their means
rentals.loc[rentals.medianServiceCharge_muni.isnull(), 'medianServiceCharge_muni'] = rentals.medianServiceCharge_muni.mean()
rentals.loc[rentals.medianServiceCharge_zip.isnull(), 'medianServiceCharge_zip'] = rentals.medianServiceCharge_zip.mean()

In [9]:
# We will drop description and facilities for now
rentals.drop(columns = ['description', 'facilities'], inplace = True)

In [10]:
#Verify that there are no missing values
rentals.isnull().any().sum()

0

# Application set development

An application set will be used to observe the performance of the model. A random sample of 5 properties will be selected. We will first of all obtain the index and location variables of this set prior to further pre-processor.

In [11]:
#Create application set of observations
application_set = rentals.sample(5, random_state = 77)
application_set

state  serviceCharge newlyConst balcony hasKitchen  \
263051         Brandenburg          160.0        Yes     Yes         No   
167352  Schleswig_Holstein          130.0         No      No        Yes   
44685              Sachsen           75.0         No     Yes        Yes   
27092       Sachsen_Anhalt           60.9         No     Yes         No   
111161              Hessen          180.0         No     Yes        Yes   

        baseRent  livingSpace                           condition  \
263051     775.0        77.70                      first_time_use   
167352     550.0        75.00  first_time_use_after_refurbishment   
44685      250.0        73.00                         refurbished   
27092      280.1        56.02                          negotiable   
111161     670.0        78.00                           well_kept   

       interiorQual lift  ... medianServiceCharge_ct  medianThermalChar_ct  \
263051      unknown  Yes  ...                  100.0                 96.00   
167352      unknown   No  ...                  115.0                135.75   
44685       unknown   No  ...                  120.0                100.00   
27092       unknown   No  ...                   92.0                104.00   
111161       normal  Yes  ...                  195.0                111.70   

        medianPictureCount_ct medianNoRooms_muni medianLivingSpace_muni  \
263051                    5.0                3.0                  60.80   
167352                    7.0                3.0                  64.64   
44685                    10.0                2.0                  62.88   
27092                     9.0                3.0                  56.98   
111161                   10.0                3.0                  83.00   

        medianServiceCharge_muni  medianPictureCount_muni  medianNoRooms_zip  \
263051                    125.00                      7.0                3.0   
167352                     95.00                      5.0                3.0   
44685                     128.00                     10.0                2.0   
27092                      69.52                      9.0                3.0   
111161                    200.00                     10.0                3.0   

        medianLivingSpace_zip  medianServiceCharge_zip  
263051                  60.80                    124.0  
167352                  64.84                     95.0  
44685                   61.50                    125.0  
27092                   56.98                     70.0  
111161                  82.50                    200.0  

[5 rows x 34 columns]

In [12]:
#Extracting the application-set location and quality features
application_set_location = application_set[['state', 'city_town', 'municipality', 'zip_code', 'condition', 'interiorQual']]
#Extracting the application-set index
application_set_index = application_set.index

# Encode categorical variables

We will using a dummy encoder to encode categorical variables. But we have to first of all drop those features with a lot of categories. These include city / town, municipality and zip code. We have already obtained relevant information for these features related to such information as to their median livingSpace. 

In [13]:
#Drop categorical features with many unique values
rentals.drop(columns = ['city_town', 'municipality', 'zip_code'], inplace = True)

In [14]:
rentals.columns

Index(['state', 'serviceCharge', 'newlyConst', 'balcony', 'hasKitchen',
       'baseRent', 'livingSpace', 'condition', 'interiorQual', 'lift',
       'typeOfFlat', 'noRooms', 'gdp_per_capita_2018', 'hdi_2018',
       'medianLivingSpace_state', 'medianServiceCharge_state',
       'medianThermalChar_state', 'medianHeatingCosts_state',
       'medianPictureCount_state', 'no_Listings_per_100people',
       'medianLivingSpace_ct', 'medianServiceCharge_ct',
       'medianThermalChar_ct', 'medianPictureCount_ct', 'medianNoRooms_muni',
       'medianLivingSpace_muni', 'medianServiceCharge_muni',
       'medianPictureCount_muni', 'medianNoRooms_zip', 'medianLivingSpace_zip',
       'medianServiceCharge_zip'],
      dtype='object')

In [15]:
#Encode cateogrical features
rentals_df = pd.get_dummies(rentals)

In [16]:
#Extracting application set samples from encoded rentals dataframe
application_df = rentals_df.loc[application_set_index]
#Drop values in the application set from dataframe
rental_df = rentals_df.drop(application_set_index)

In [17]:
rentals_df.shape

(267670, 74)

In [18]:
application_df.shape

(5, 74)

# Create a Dataframe for the features and target variable

In [19]:
X = rentals_df.drop(columns = ['baseRent'])
y = rentals_df.baseRent
X_features = X.columns

In [20]:
X_features

Index(['serviceCharge', 'livingSpace', 'noRooms', 'gdp_per_capita_2018',
       'hdi_2018', 'medianLivingSpace_state', 'medianServiceCharge_state',
       'medianThermalChar_state', 'medianHeatingCosts_state',
       'medianPictureCount_state', 'no_Listings_per_100people',
       'medianLivingSpace_ct', 'medianServiceCharge_ct',
       'medianThermalChar_ct', 'medianPictureCount_ct', 'medianNoRooms_muni',
       'medianLivingSpace_muni', 'medianServiceCharge_muni',
       'medianPictureCount_muni', 'medianNoRooms_zip', 'medianLivingSpace_zip',
       'medianServiceCharge_zip', 'state_Baden_Württemberg', 'state_Bayern',
       'state_Berlin', 'state_Brandenburg', 'state_Bremen', 'state_Hamburg',
       'state_Hessen', 'state_Mecklenburg_Vorpommern', 'state_Niedersachsen',
       'state_Nordrhein_Westfalen', 'state_Rheinland_Pfalz', 'state_Saarland',
       'state_Sachsen', 'state_Sachsen_Anhalt', 'state_Schleswig_Holstein',
       'state_Thüringen', 'newlyConst_No', 'newlyConst_Yes', 'b

# Scaling the features

In [21]:
scaler = StandardScaler(with_mean = False)
X_scaled = scaler.fit_transform(X)

In [22]:
X_scaled[:1]

array([[ 2.76544573,  2.677037  ,  4.07261169,  5.32527978, 68.61715701,
         9.77600442,  5.94304803, 10.88519767, 13.23631072, 11.63621134,
         0.72907196,  7.72822371,  3.92930073,  7.48333364,  5.89452368,
         5.94717572,  6.29471578,  4.08542212,  4.02837852,  6.00313939,
         6.18224529,  3.62615569,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  2.36118706,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ,  3.72042691,  0.        ,
         2.05667054,  0.        ,  2.10777964,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ,  0.        ,  2.31624267,
         0.        ,  2.1723355 ,  0.        ,  0.        ,  0.        ,
         2.35798375,  0.        ,  0.        ,  3.10691003,  0.        ,
         0.        ,  0.        ,  0.        ,  0. 

# Create Train-Test Split

In [23]:
#Split X and y to test and training set
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 77)

# Baseline model

We will be using a linear regression algorithm to model the relationship between base rent and the features. We will also use the r-squared score and the mean absolute error to assess the model's performance. The r-squared score gives us an indication as to how much of the variation of base rent is due to the model, while the mean absolute error gives us a realistic sense for the margin of error associated with the model in terms of euros. 

In [24]:
#Instantiate linear regression model
lm = LinearRegression()

In [25]:
#Fit the model to the data
lm.fit(X_train, y_train);

In [26]:
#Determine predictions for train and the test set
y_train_pred = lm.predict(X_train)
y_test_pred = lm.predict(X_test)

In [27]:
#Review performance of model
r2_score(y_train, y_train_pred), mean_absolute_error(y_train, y_train_pred)

(0.7264881063744089, 146.28933067903444)

In [28]:
#Review performance of model
r2_score(y_test, y_test_pred), mean_absolute_error(y_test, y_test_pred)

(0.7478734650432224, 147.22967845979502)

# Summary

After processing the data to enable modelling, our base model gives us a 72% r2-score and a 147 euros mean absolute error on our test set. It seems like our model is informative and could be improved upond with other algorithms and modelling techniques. Our training set scores were 73.5% for r2-score 147 euros for mae. The closeness of set of values for our training and test sets suggests that there is minimal overfitting of the model. 

Going forward, we will attempt to improve on these scores by:
- Choosing select features
- Using other modelling algorithms
- Tuning hyperparameters

# Saving

In [29]:
import pickle
filepath = '../data/interim/'
pickle.dump(lm, open(filepath + 'lm_base.pkl', 'wb'))
pickle.dump(X_test, open(filepath + 'X_test_base.pkl', 'wb'))
pickle.dump(y_test, open(filepath + 'y_test_base.pkl', 'wb'))
pickle.dump(X_train, open(filepath + 'X_train_base.pkl', 'wb'))
pickle.dump(y_train, open(filepath + 'y_train_base.pkl', 'wb'))
pickle.dump(X_features, open(filepath + 'X_features_base.pkl', 'wb'))
pickle.dump(X_scaled, open(filepath + 'X_scaled.pkl', 'wb'))
pickle.dump(y, open(filepath + 'y.pkl', 'wb'))
pickle.dump(application_df, open(filepath + 'application_df.pkl', 'wb'))
pickle.dump(application_set_location, open(filepath + 'application_set_location.pkl', 'wb'))